File: src/main/providers/google-oauth.ts
import { constants } from "node:fs";
import { access, readdir, readFile, realpath } from "node:fs/promises";
import { dirname, join } from "node:path";
import { resolveExecutable } from "./common";

const OAUTH_SOURCE_PATH = join("dist", "src", "code_assist", "oauth2.js");

async function exists(path: string): Promise<boolean> {
  try {
    await access(path, constants.R_OK);
    return true;
  } catch {
    return false;
  }
}

function parseClientCredentials(
  source: string,
): { clientId: string; clientSecret: string } | null {
  const clientId = source.match(/OAUTH_CLIENT_ID\s*=\s*["']([^"']+)["']/)?.[1];
  const clientSecret = source.match(
    /OAUTH_CLIENT_SECRET\s*=\s*["']([^"']+)["']/,
  )?.[1];
  return clientId && clientSecret ? { clientId, clientSecret } : null;
}

async function readClientCredentials(
  path: string,
): Promise<{ clientId: string; clientSecret: string } | null> {
  try {
    return parseClientCredentials(await readFile(path, "utf8"));
  } catch {
    return null;
  }
}

async function packageRoot(binary: string): Promise<string | null> {
  let current = dirname(binary);
  for (let depth = 0; depth <= 8; depth += 1) {
    try {
      const manifest = JSON.parse(
        await readFile(join(current, "package.json"), "utf8"),
      ) as { name?: unknown };
      if (manifest.name === "@google/gemini-cli") return current;
    } catch {
      // Global installs often put the launcher several levels above the package.
    }
    for (const nested of [
      join(current, "lib", "node_modules", "@google", "gemini-cli"),
      join(current, "node_modules", "@google", "gemini-cli"),
    ]) {
      if (await exists(join(nested, "package.json"))) return nested;
    }
    const parent = dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return null;
}

export async function extractGeminiOAuthClient(
  configuredExecutable: string | null,
): Promise<{ clientId: string; clientSecret: string } | null> {
  const executable = await resolveExecutable("gemini", configuredExecutable);
  if (!executable) return null;
  const resolved = await realpath(executable).catch(() => executable);
  const binDirectory = dirname(resolved);
  const prefix = dirname(binDirectory);
  const candidates = [
    join(
      prefix,
      "libexec",
      "lib",
      "node_modules",
      "@google",
      "gemini-cli",
      "node_modules",
      "@google",
      "gemini-cli-core",
      OAUTH_SOURCE_PATH,
    ),
    join(
      prefix,
      "lib",
      "node_modules",
      "@google",
      "gemini-cli",
      "node_modules",
      "@google",
      "gemini-cli-core",
      OAUTH_SOURCE_PATH,
    ),
    join(prefix, "..", "gemini-cli-core", OAUTH_SOURCE_PATH),
    join(
      prefix,
      "node_modules",
      "@google",
      "gemini-cli-core",
      OAUTH_SOURCE_PATH,
    ),
  ];
  for (const candidate of candidates) {
    const credentials = await readClientCredentials(candidate);
    if (credentials) return credentials;
  }

  const root = await packageRoot(resolved);
  if (!root) return null;
  for (const candidate of [
    join(root, "node_modules", "@google", "gemini-cli-core", OAUTH_SOURCE_PATH),
    join(root, OAUTH_SOURCE_PATH),
  ]) {
    const credentials = await readClientCredentials(candidate);
    if (credentials) return credentials;
  }

  const bundle = join(root, "bundle");
  const entries = await readdir(bundle).catch(() => []);
  for (const entry of entries) {
    if (!entry.endsWith(".js")) continue;
    const credentials = await readClientCredentials(join(bundle, entry));
    if (credentials) return credentials;
  }
  return null;
}

File: src-tauri/core/src/providers/google_oauth.rs
use super::common::resolve_executable;
use regex::Regex;
use serde_json::Value;
use std::{
    fs,
    path::{Path, PathBuf},
    sync::LazyLock,
};

fn parse(source: &str) -> Option<(String, String)> {
    static ID: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r#"OAUTH_CLIENT_ID\s*=\s*["']([^"']+)["']"#).unwrap());
    static SECRET: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r#"OAUTH_CLIENT_SECRET\s*=\s*["']([^"']+)["']"#).unwrap());
    Some((
        ID.captures(source)?.get(1)?.as_str().into(),
        SECRET.captures(source)?.get(1)?.as_str().into(),
    ))
}

fn read(path: &Path) -> Option<(String, String)> {
    parse(&fs::read_to_string(path).ok()?)
}

fn package_root(binary: &Path) -> Option<PathBuf> {
    let mut current = binary.parent()?.to_path_buf();
    for _ in 0..=8 {
        if fs::read(current.join("package.json"))
            .ok()
            .and_then(|data| serde_json::from_slice::<Value>(&data).ok())
            .and_then(|value| value.get("name").and_then(Value::as_str).map(str::to_owned))
            .as_deref()
            == Some("@google/gemini-cli")
        {
            return Some(current);
        }
        for nested in [
            current.join("lib/node_modules/@google/gemini-cli"),
            current.join("node_modules/@google/gemini-cli"),
        ] {
            if nested.join("package.json").is_file() {
                return Some(nested);
            }
        }
        if !current.pop() {
            break;
        }
    }
    None
}

pub async fn extract(configured: Option<&Path>) -> Option<(String, String)> {
    let binary = resolve_executable("gemini", configured).await?;
    let resolved = fs::canonicalize(&binary).unwrap_or(binary);
    let bin = resolved.parent()?;
    let prefix = bin.parent()?;
    const SOURCE: &str = "dist/src/code_assist/oauth2.js";
    for candidate in [
        prefix
            .join(
                "libexec/lib/node_modules/@google/gemini-cli/node_modules/@google/gemini-cli-core",
            )
            .join(SOURCE),
        prefix
            .join("lib/node_modules/@google/gemini-cli/node_modules/@google/gemini-cli-core")
            .join(SOURCE),
        prefix.join("../gemini-cli-core").join(SOURCE),
        prefix
            .join("node_modules/@google/gemini-cli-core")
            .join(SOURCE),
    ] {
        if let Some(value) = read(&candidate) {
            return Some(value);
        }
    }
    let root = package_root(&resolved)?;
    for candidate in [
        root.join("node_modules/@google/gemini-cli-core")
            .join(SOURCE),
        root.join(SOURCE),
    ] {
        if let Some(value) = read(&candidate) {
            return Some(value);
        }
    }
    for entry in fs::read_dir(root.join("bundle")).ok()? {
        let path = entry.ok()?.path();
        if path.extension().and_then(|value| value.to_str()) == Some("js") {
            if let Some(value) = read(&path) {
                return Some(value);
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    #[test]
    fn parses_client_constants() {
        assert_eq!(
            super::parse("const OAUTH_CLIENT_ID = 'id'; const OAUTH_CLIENT_SECRET=\"secret\";"),
            Some(("id".into(), "secret".into()))
        );
    }
}
